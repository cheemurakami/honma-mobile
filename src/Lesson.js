import { Text, View } from "react-native";

import Icon from "react-native-vector-icons/FontAwesome";
import React from "react";
import { Title } from 'react-native-paper';

const Lesson = ({ navigation }) => {
  return (
    <View>
      <Icon
        name="arrow-left"
        size={30}
        onPress={() => navigation.navigate("ChooseDialect")}
      />
      <Title style={{textAlign: 'center'}}>いる / おる　おるん?</Title>
    </View>
  );
};

export default Lesson;
