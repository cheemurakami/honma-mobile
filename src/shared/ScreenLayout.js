import FooterButton from "../shared/FooterButton";
import Icon from "react-native-vector-icons/FontAwesome";
import React from "react";
import { Title } from "react-native-paper";
import { View } from "react-native";
import styled from "styled-components/native";
import { useNavigation } from "@react-navigation/native";

const ScreenLayout = (props) => {
  const navigation = useNavigation();

  const {
    pageTitle,
    btnLabel,
    backComponentName,
    children,
    onPressHandler,
  } = props;

  return (
    <View>
      <Icon
        name="arrow-left"
        size={30}
        style={{ margin: 10 }}
        onPress={() => navigation.navigate(backComponentName)}
      />
      <Title style={{ textAlign: "center" }}>{pageTitle}</Title>
      <BodyContainer>{children}</BodyContainer>
      <Footer>
        <FooterButton title={btnLabel} onPressHandler={onPressHandler} />
      </Footer>
    </View>
  );
};

const BodyContainer = styled.View`
  background-color: white;
  height: 80%;
  margin: 10px;
`;

const Footer = styled.View`
  align-items: center;
`;

export default ScreenLayout;
