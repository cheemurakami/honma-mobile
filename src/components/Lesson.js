import * as a from "../rdx/actions";

import React, { useState } from "react";

import { Button } from "react-native-paper";
import Commonness from "./Commonness";
import ScreenLayout from "../shared/ScreenLayout";
import SoundPlayButton from "../shared/SoundPlayButton";
import { TextInput } from "react-native-paper";
import { connect } from "react-redux";
import styled from "styled-components/native";

const Lesson = ({ route, navigation, dispatch, completedGrammars }) => {
  const [text, setText] = useState("");
  const { selectedDialect, grammar } = route.params;
  const btnLabel = selectedDialect.complete_btn_text;
  const jpExample = grammar.examples.find(
    (example) => example.language === "jp"
  );
  const enExample = grammar.examples.find(
    (example) => example.language === "en"
  );

  const completeBtn = () => {
    const action = a.completedGrammars(grammar.id);
    dispatch(action);
    const nextGrammar = selectedDialect.grammars.find(
      (g) => g.position === grammar.position + 1
    );
    if (nextGrammar) {
      const action = a.selectedGrammar(nextGrammar.id);
      dispatch(action);
      navigation.navigate("Lesson", { selectedDialect, grammar: nextGrammar });
    } else {
      navigation.navigate("PatternList", { selectedDialect });
    }
  };

  const showCompletedAt = () => {
    if (completedGrammars[grammar.id]) {
      const completedDate = new Date(
        completedGrammars[grammar.id]
      ).toLocaleDateString();
      return <BodySubText>Completed at {completedDate}</BodySubText>;
    }
  };

  return (
    <ScreenLayout
      pageTitle={grammar.label}
      btnLabel={btnLabel}
      backComponentName={"PatternList"}
      onPressHandler={() => completeBtn()}
    >
      <BodyText>{grammar.description}</BodyText>
      <Commonness commonness={grammar.commonness} />
      {showCompletedAt()}
      <MediaContainer>
        <ExampleContainer>
          <BodyTextExample>A: {jpExample.sentence1}</BodyTextExample>
          <BodyTextExample>B: {jpExample.sentence2}</BodyTextExample>
        </ExampleContainer>

        {jpExample.audio_clip_url && (
          <SoundPlayButton soundSource={jpExample.audio_clip_url} />
        )}
      </MediaContainer>
      <ExampleContainer>
        <BodyTextExample>A: {enExample.sentence1}</BodyTextExample>
        <BodyTextExample>B: {enExample.sentence2}</BodyTextExample>
      </ExampleContainer>

      <BodyText>Please write this in {selectedDialect.name_en}:</BodyText>
      <BodyText>お母さんは部屋にいるよ。</BodyText>
      <TextInput
        label="Answer here"
        value={text}
        onChangeText={(text) => setText(text)}
        style={{ margin: 20 }}
      ></TextInput>
      <ButtonContainer>
        <Button
          mode="contained"
          onPress={() => console.log("button pressed")}
          style={{
            fontSize: 16,
            color: "#fff",
            backgroundColor: "#40BA62",
            width: "50%",
            height: 45,
            justifyContent: "center",
            alignItems: "center",
          }}
        >
          答え合わせ
        </Button>
      </ButtonContainer>
    </ScreenLayout>
  );
};

const BodyText = styled.Text`
  text-align: center;
  font-size: 18px;
  margin: 5px;
  font-weight: bold;
`;

const BodySubText = styled.Text`
  margin-left: 15px;
  font-size: 16px;
  font-weight: bold;
  color: #a3a3a3;
`;

const BodyTextExample = styled.Text`
  text-align: left;
  font-size: 20px;
  margin: 10px;
`;

const MediaContainer = styled.View`
  flex-direction: row;
  align-items: center;
`;

const ExampleContainer = styled.View`
  padding: 5px;
  margin-top: 10px;
`;
const ButtonContainer = styled.View`
  align-items: center;
`;

const mapStateToProps = (state) => {
  return {
    completedGrammars: state.grammarsReducer.completedIds,
  };
};
export default connect(mapStateToProps)(Lesson);
