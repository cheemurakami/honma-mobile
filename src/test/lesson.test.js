import React from "react";
import renderer from "react-test-renderer";
import { NavigationContainer } from "@react-navigation/native";
import { configureStore } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import rootReducer from "../rdx/reducers/index.js";
import Lesson from "../components/Lesson.js";
import ScreenLayout from "../shared/ScreenLayout.js";
import { waitFor, render, screen } from "@testing-library/react-native";
import {
  dialectData,
  grammarWithOneQuiz,
  grammarWithMultipleQuiz,
  grammarWithCompletedMultipleQuiz,
} from "./test-data/lesson-test-data.js";

const mockDispatch = jest.fn();

jest.mock("@react-navigation/native", () => {
  const actualNav = jest.requireActual("@react-navigation/native");
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: jest.fn(),
      dispatch: mockDispatch,
    }),
  };
});

jest.mock("../components/Modal", () => ({
  ...jest.requireActual("../components/Modal"),
  FindById: jest.fn(() => "mocked response"),
}));

jest.mock("../shared/ScreenLayout.js", () => {
  const React = require("react");
  const ScreenLayout = ({ children }) => {
    return <div data-testid="mocked-screen-layout">{children}</div>;
  };
  return ScreenLayout;
});

// KeyboardAwareScrollView causes the error: "Can't access .root on unmounted test renderer." This mock is needed to prevent this error.
jest.mock("react-native-keyboard-aware-scroll-view", () => {
  const KeyboardAwareScrollView = require("react-native").ScrollView;
  return { KeyboardAwareScrollView };
});

describe("Lesson tests", () => {
  const store = configureStore({ reducer: rootReducer });

  const testRoute1 = {
    key: "Lesson-CpYLFUxzpmxStg-7Ti3tQ",
    name: "Lesson",
    params: {
      allQuizzesCompleted: true,
      grammar: grammarWithOneQuiz,
      selectedDialect: dialectData,
    },
  };

  const testRoute2 = {
    key: "Lesson-CpYLFUxzpmxStg-7Ti3tQ",
    name: "Lesson",
    params: {
      allQuizzesCompleted: false,
      grammar: grammarWithMultipleQuiz,
      selectedDialect: dialectData,
    },
  };

  const testRoute3 = {
    key: "Lesson-CpYLFUxzpmxStg-7Ti3tQ",
    name: "Lesson",
    params: {
      allQuizzesCompleted: true,
      grammar: grammarWithCompletedMultipleQuiz,
      selectedDialect: dialectData,
    },
  };

  const testNavigation = {
    navigate: jest.fn(),
  };

  const renderComponent = (quizCount) => {
    let testRoute;
    if (quizCount === "one quiz") {
      testRoute = testRoute1;
    } else if (quizCount === "multiple quizzes") {
      testRoute = testRoute2;
    } else {
      testRoute = testRoute3;
    }

    render(
      <Provider store={store}>
        <NavigationContainer>
          <ScreenLayout>
            <Lesson
              route={testRoute}
              navigation={testNavigation}
              dispatch={mockDispatch}
            />
          </ScreenLayout>
        </NavigationContainer>
      </Provider>
    );
  };

  it("matches snapshot", async () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <NavigationContainer>
            <ScreenLayout>
              <Lesson
                route={testRoute1}
                navigation={testNavigation}
                dispatch={mockDispatch}
              />
            </ScreenLayout>
          </NavigationContainer>
        </Provider>
      )
      .toJSON();
    await waitFor(() => {
      expect(tree).toMatchSnapshot();
    });
  });

  it("shows completed at date when all the quizzes are completed", async () => {
    renderComponent("one quiz");

    // const date = grammarWithOneQuiz.quizzes[0].quiz_completed;
    // const formattedDate = new Date(date).toLocaleDateString();
    const formattedDate = "9/7/2023";
    const completedAtText = screen.getByText(`Completed at ${formattedDate}`);

    await waitFor(() => {
      expect(completedAtText).toBeTruthy();
    });
  });

  it("shows most completed at date when all the quizzes are completed", async () => {
    renderComponent("completed all quizzes");

    const formattedDate = "9/30/2023";
    const completedAtText = screen.getByText(`Completed at ${formattedDate}`);

    await waitFor(() => {
      expect(completedAtText).toBeTruthy();
    });
  });

  it("does not show completed at date when all the quizzes are not completed", async () => {
    renderComponent("multiple quizzes");

    const completedAtText = screen.queryByText("Completed at");
    await waitFor(() => {
      expect(completedAtText).not.toBeTruthy();
    });
  });

  describe('buttons', () => {
    it("shows proper buttons when all the quizzes are completed", async () => {
      renderComponent("one quiz")

      const quizButton = screen.getByLabelText("quiz-button")
      const resetButton = screen.getByLabelText("reset-quizzes-button")
      const quizButtonText = screen.getByText("正解したクイズを見る")
      const startQuizText = screen.queryByText("クイズに挑戦🌟")

      await waitFor(() => {
        expect(quizButton).toBeTruthy();
        expect(resetButton).toBeTruthy();
        expect(quizButtonText).toBeTruthy();
        expect(startQuizText).not.toBeTruthy();
      });
    });

    it("shows proper button when the quizzes are not completed", async () => {
      renderComponent("multiple quizzes")

      const quizButton = screen.getByLabelText("quiz-button")
      const resetButton = screen.queryByLabelText("reset-quizzes-button")
      const quizButtonText = screen.queryByText("正解したクイズを見る")
      const startQuizText = screen.getByText("クイズに挑戦🌟")


      await waitFor(() => {
        expect(quizButton).toBeTruthy();
        expect(resetButton).not.toBeTruthy();
        expect(quizButtonText).not.toBeTruthy();
        expect(startQuizText).toBeTruthy();
      });
    });
  })
});
