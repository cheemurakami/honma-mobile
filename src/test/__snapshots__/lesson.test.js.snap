// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lesson tests matches snapshot 1`] = `
<div
  data-testid="mocked-screen-layout"
>
  <div
    data-testid="mocked-screen-layout"
  >
    <RCTScrollView>
      <View>
        <RCTScrollView
          contentContainerStyle={
            {
              "flex": 1,
            }
          }
          resetScrollToCoords={
            {
              "x": 0,
              "y": 0,
            }
          }
          scrollEnabled={true}
        >
          <View>
            <Text
              style={
                [
                  {
                    "fontSize": 18,
                    "fontWeight": "bold",
                    "marginBottom": 5,
                    "marginLeft": 5,
                    "marginRight": 5,
                    "marginTop": 5,
                    "textAlign": "center",
                  },
                ]
              }
            >
              よ changes to なんよ
            </Text>
            <Text
              style={
                [
                  {
                    "fontSize": 12,
                    "fontWeight": "bold",
                    "marginBottom": 15,
                    "marginLeft": 15,
                    "marginRight": 15,
                    "marginTop": 15,
                  },
                ]
              }
            >
              Commonness: 
            </Text>
            <Text
              style={
                [
                  {
                    "color": "#a3a3a3",
                    "fontSize": 16,
                    "fontWeight": "bold",
                    "marginLeft": 15,
                  },
                ]
              }
            >
              Completed at 
              9/7/2023
            </Text>
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "flexDirection": "row",
                  },
                ]
              }
            >
              <View
                style={
                  [
                    {
                      "marginTop": 10,
                    },
                  ]
                }
              >
                <Text
                  style={
                    [
                      {
                        "fontSize": 20,
                        "marginBottom": 10,
                        "marginLeft": 10,
                        "marginRight": 10,
                        "marginTop": 10,
                        "textAlign": "left",
                      },
                    ]
                  }
                >
                  A: 
                  車、買ったん?
                </Text>
                <Text
                  style={
                    [
                      {
                        "fontSize": 20,
                        "marginBottom": 10,
                        "marginLeft": 10,
                        "marginRight": 10,
                        "marginTop": 10,
                        "textAlign": "left",
                      },
                    ]
                  }
                >
                  B: 
                  そうなんよ。
                </Text>
              </View>
            </View>
            <View
              style={
                [
                  {
                    "marginTop": 10,
                  },
                ]
              }
            >
              <Text
                style={
                  [
                    {
                      "fontSize": 20,
                      "marginBottom": 10,
                      "marginLeft": 10,
                      "marginRight": 10,
                      "marginTop": 10,
                      "textAlign": "left",
                    },
                  ]
                }
              >
                A: 
                Did you buy a car?
              </Text>
              <Text
                style={
                  [
                    {
                      "fontSize": 20,
                      "marginBottom": 10,
                      "marginLeft": 10,
                      "marginRight": 10,
                      "marginTop": 10,
                      "textAlign": "left",
                    },
                  ]
                }
              >
                B: 
                Yes, I did.
              </Text>
            </View>
            <View
              style={
                [
                  {
                    "alignItems": "center",
                    "marginBottom": 10,
                    "marginLeft": 10,
                    "marginRight": 10,
                    "marginTop": 10,
                  },
                ]
              }
            >
              <View
                collapsable={false}
                style={
                  {
                    "backgroundColor": "#1B70B1",
                    "borderColor": "transparent",
                    "borderRadius": 4,
                    "borderStyle": "solid",
                    "borderWidth": 0,
                    "elevation": 2,
                    "height": 45,
                    "justifyContent": "center",
                    "minWidth": 64,
                    "shadowColor": "#000000",
                    "shadowOffset": {
                      "height": 0.75,
                      "width": 0,
                    },
                    "shadowOpacity": 0.24,
                    "shadowRadius": 1.5,
                    "width": "60%",
                  }
                }
              >
                <View
                  accessibilityRole="button"
                  accessibilityState={
                    {
                      "disabled": false,
                    }
                  }
                  accessibilityValue={
                    {
                      "max": undefined,
                      "min": undefined,
                      "now": undefined,
                      "text": undefined,
                    }
                  }
                  accessible={true}
                  focusable={true}
                  onClick={[Function]}
                  onResponderGrant={[Function]}
                  onResponderMove={[Function]}
                  onResponderRelease={[Function]}
                  onResponderTerminate={[Function]}
                  onResponderTerminationRequest={[Function]}
                  onStartShouldSetResponder={[Function]}
                  style={
                    [
                      {
                        "overflow": "hidden",
                      },
                      {
                        "borderRadius": 4,
                      },
                    ]
                  }
                >
                  <View
                    style={
                      [
                        {
                          "alignItems": "center",
                          "flexDirection": "row",
                          "justifyContent": "center",
                        },
                        undefined,
                      ]
                    }
                  >
                    <Text
                      numberOfLines={1}
                      selectable={false}
                      style={
                        [
                          {
                            "color": "#000000",
                            "fontFamily": "System",
                            "fontWeight": "400",
                          },
                          {
                            "textAlign": "left",
                          },
                          [
                            {
                              "letterSpacing": 1,
                              "marginHorizontal": 16,
                              "marginVertical": 9,
                              "textAlign": "center",
                            },
                            undefined,
                            {
                              "textTransform": "uppercase",
                            },
                            {
                              "color": "#ffffff",
                              "fontFamily": "System",
                              "fontWeight": "500",
                            },
                            {
                              "fontFamily": "System",
                              "fontWeight": "500",
                            },
                            {
                              "color": "#fff",
                              "fontSize": 18,
                            },
                          ],
                        ]
                      }
                    >
                      正解したクイズを見る
                    </Text>
                  </View>
                </View>
              </View>
            </View>
            <Image
              source={1}
              style={
                {
                  "height": 100,
                  "width": 100,
                }
              }
            />
          </View>
        </RCTScrollView>
      </View>
    </RCTScrollView>
  </div>
</div>
`;
